packages/provider/src/MultiChainProvider.test.ts

```diff
@@ -0,0 +1,404 @@
+        jsonrpc: '2.0',
+        id: 1,
+        result: {
+          namespaces: {
+            eip155: 'foo',
+          },
+        },
+      }));
+
+      const provider = new MultiChainProvider();
+      expect(provider.isConnected).toBe(false);
+
+      const listener = jest.fn();
+      provider.on('session_update', listener);
+
+      const { approval } = await provider.connect({ requiredNamespaces: {} });
+      await expect(approval()).rejects.toThrow(
+        'Invalid session: At path: namespaces.eip155 -- Expected an object, but received: "foo"',
+      );
+
+      expect(provider.isConnected).toBe(false);
+      expect(listener).not.toHaveBeenCalled();
+    });
+
+    it('throws on JSON-RPC error response', async () => {
+      const request = ethereum.request as jest.MockedFunction<
+        typeof ethereum.request
+      >;
+
+      request.mockImplementation(async () => ({
+        jsonrpc: '2.0',
+        id: 1,
+        error: {
+          code: -1,
+          message: 'foo',
+        },
+      }));
+
+      const provider = new MultiChainProvider();
+      expect(provider.isConnected).toBe(false);
+
+      const { approval } = await provider.connect({ requiredNamespaces: {} });
+      await expect(approval()).rejects.toThrow(
+        'Not a successful JSON-RPC response: At path: result -- Expected the value to satisfy a union of `literal | boolean | number | string | lazy | lazy`, but received: undefined.',
+      );
+    });
+  });
+
+  describe('request', () => {
+    it('throws if not connected', async () => {
+      const provider = new MultiChainProvider();
+      expect(provider.isConnected).toBe(false);
+
+      await expect(
+        provider.request({
+          chainId: 'eip155:1',
+          request: {
+            method: 'eth_accounts',
+          },
+        }),
+      ).rejects.toThrow('No session connected.');
+    });
+
+    it.each([
+      null,
+      undefined,
+      {
+        chainId: 'eip155:1',
+        request: 'foo',
+      },
+      {
+        chainId: 'foo',
+        request: {
+          method: 'eth_accounts',
+        },
+      },
+    ])('validates the request arguments', async (args) => {
+      const provider = await getProvider();
+
+      // @ts-expect-error Invalid arguments.
+      await expect(provider.request(args)).rejects.toThrow(
+        'Invalid request arguments',
+      );
+    });
+
+    it('returns the response result', async () => {
+      const provider = await getProvider();
+
+      const request = ethereum.request as jest.MockedFunction<
+        typeof ethereum.request
+      >;
+
+      request.mockImplementation(async () => ({
+        jsonrpc: '2.0',
+        id: 1,
+        result: 'foo',
+      }));
+
+      const result = await provider.request({
+        chainId: 'eip155:1',
+        request: {
+          method: 'eth_accounts',
+        },
+      });
+
+      expect(result).toBe('foo');
+    });
+
+    it('generates a random request ID', async () => {
```

Comment:
A good addition to this test would be to do 2 requests and check if ids are different